File: backend/java/rf-core/src/main/java/com/rfindustries/core/service/BaseCrudService.java
package com.rfindustries.core.service;

import com.rfindustries.core.beans.ResponseMethod;
import com.rfindustries.core.dao.BaseDao;
import com.rfindustries.core.dto.BaseDTO;
import com.rfindustries.core.entities.BaseEntity;
import com.rfindustries.core.features.BaseCommonsParameters;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.List;

public interface BaseCrudService<DAO extends BaseDao<ENTITY, PK>, ENTITY extends BaseEntity<PK>, PK, DTO extends BaseDTO> extends BaseService {

    DAO getDao();

    ResponseMethod<Flux<DTO>> insertAll(BaseCommonsParameters baseCommonsParameters, List<DTO> dtos);

    ResponseMethod<Mono<DTO>> insert(BaseCommonsParameters baseCommonsParameters, DTO dto);

    ResponseMethod<Flux<DTO>> updateAll(BaseCommonsParameters baseCommonsParameters, List<DTO> dtos);

    ResponseMethod<Mono<DTO>> update(BaseCommonsParameters baseCommonsParameters, DTO dto);

    ENTITY toEntity(BaseCommonsParameters baseCommonsParameters, DTO dto);

    DTO toDTO(BaseCommonsParameters baseCommonsParameters, ENTITY entity);

    DTO instanceDTO();

    default ResponseMethod<DTO> actionDoAfterInsertUpdate(BaseCommonsParameters baseCommonsParameters, DTO dto, boolean insert) {
        return ResponseMethod.<DTO>builder().data(dto).build();
    }

    default ResponseMethod<DTO> actionDoBeforeInsertUpdate(BaseCommonsParameters baseCommonsParameters, DTO dto, boolean insert) {
        return ResponseMethod.<DTO>builder().data(dto).build();
    }

    ResponseMethod<Mono<Boolean>> delete(BaseCommonsParameters baseCommonsParameters, DTO dto);


    default ResponseMethod<DTO> actionDoBeforeDelete(BaseCommonsParameters baseCommonsParameters, DTO dto) {
        return ResponseMethod.<DTO>builder().data(dto).build();
    }

    default ResponseMethod<DTO> actionDoAfterDelete(BaseCommonsParameters baseCommonsParameters, DTO dto) {
        return ResponseMethod.<DTO>builder().data(dto).build();
    }

    Mono<DTO> findById(BaseCommonsParameters baseCommonsParameters, PK pk);

    default <HEADER_PK> Flux<DTO> findByHeaderPk(BaseCommonsParameters baseCommonsParameters, HEADER_PK headerPk) {
        return Flux.empty();
    }

    ResponseMethod<Mono<DTO>> goRead(BaseCommonsParameters baseCommonsParameters, PK pk);

    ResponseMethod<Mono<DTO>> goEdit(BaseCommonsParameters baseCommonsParameters, PK pk);

    ResponseMethod<Mono<DTO>> goAdd(BaseCommonsParameters baseCommonsParameters);

    default <HEADER_PK> Mono<Long> deleteByHeaderPk(BaseCommonsParameters baseCommonsParameters, HEADER_PK headerPk) {
        return Mono.just(0L);
    }

    default PK resolvePK(DTO dto) {
        return null;
    }
}

File: backend/go/rf-core/pkg/service/basecrud_service.go
package service

import (
	"rfindustries/rf-core/pkg/dao"
	"rfindustries/rf-core/pkg/features"
)

type BaseCrudService interface {
	GetDao(self BaseCrudService, commonsParameters features.BaseCommonsParamenters) *dao.BaseDao
	CheckBeforeInsertUpdate(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, dto interface{}, insert bool) (interface{}, error)
	CheckAfterInsertUpdate(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, dto interface{}, insert bool) (interface{}, error)
	Insert(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, dto interface{}) (interface{}, error)
	Update(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, dto interface{}) (interface{}, error)
	Delete(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, dto interface{}) (bool, error)
	ToDTO(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, entity interface{}) interface{}
	ToEntity(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, dto interface{}) interface{}
}

type BaseCrudServiceImpl struct {
	Dao dao.BaseDao
}

func (baseCrudService BaseCrudServiceImpl) GetDao(self BaseCrudService, commonsParameters features.BaseCommonsParamenters) dao.BaseDao {
	return baseCrudService.Dao
}

func (baseCrudService BaseCrudServiceImpl) CheckBeforeInsertUpdate(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, dto interface{}, insert bool) (interface{}, error) {
	result := dto
	var err error

	if self != nil {
		result, dto = self.CheckBeforeInsertUpdate(nil, commonsParameters, dto, insert)
	}

	return result, err
}

func (baseCrudService BaseCrudServiceImpl) CheckAfterInsertUpdate(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, dto interface{}, insert bool) (interface{}, error) {
	result := dto
	var err error

	if self != nil {
		result, dto = self.CheckAfterInsertUpdate(nil, commonsParameters, dto, insert)
	}

	return result, err
}

func (baseCrudService BaseCrudServiceImpl) Insert(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, dto interface{}) (interface{}, error) {
	dto, err := baseCrudService.CheckBeforeInsertUpdate(self, commonsParameters, dto, true)

	if err == nil {
		entity := self.ToEntity(nil, commonsParameters, dto)
		dao := baseCrudService.GetDao(nil, commonsParameters)
		entity = dao.Insert(dao, commonsParameters, entity)
		dto = self.ToDTO(self, commonsParameters, entity)
		dto, err = self.CheckAfterInsertUpdate(nil, commonsParameters, dto, true)
	}

	return dto, err
}

func (baseCrudService BaseCrudServiceImpl) Update(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, dto interface{}) (interface{}, error) {
	dto, err := baseCrudService.CheckBeforeInsertUpdate(self, commonsParameters, dto, false)

	if err == nil {
		entity := self.ToEntity(nil, commonsParameters, dto)
		dao := baseCrudService.GetDao(nil, commonsParameters)
		entity = dao.Update(dao, commonsParameters, entity)
		dto = self.ToDTO(self, commonsParameters, entity)
		dto, err = self.CheckAfterInsertUpdate(nil, commonsParameters, dto, false)
	}

	return dto, err
}

func (baseCrudService BaseCrudServiceImpl) Delete(self BaseCrudService, commonsParameters features.BaseCommonsParamenters, dto interface{}) (bool, error) {
	entity := self.ToEntity(nil, commonsParameters, dto)
	var err error

	dao := baseCrudService.GetDao(nil, commonsParameters)

	return dao.Delete(dao, commonsParameters, entity), err
}
